In [2]:
!pip install kagglehub

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.4/63.4 kB 7.6 MB/s eta 0:00:00


In [8]:
import numpy as np
import pandas as pd
import os

# Sales Data Cleaning

In [3]:
import kagglehub

path = kagglehub.dataset_download("kyanyoga/sample-sales-data")

print("Path to dataset files:", path)

100%|███████████████████████████████████████████████████████████████████████████████| 77.5k/77.5k [00:00<00:00, 162kB/s]

Extracting files...
Path to dataset files: /home/24n0457/.cache/kagglehub/datasets/kyanyoga/sample-sales-data/versions/1


In [15]:
path1 = path

In [16]:
path1

'/home/24n0457/.cache/kagglehub/datasets/kyanyoga/sample-sales-data/versions/1'

In [17]:
print(os.listdir(path1))

['sales_data_sample.csv']


In [20]:
file_path = os.path.join(path, 'sales_data_sample.csv')

In [21]:
file_path

'/home/24n0457/.cache/kagglehub/datasets/kyanyoga/sample-sales-data/versions/1/sales_data_sample.csv'

In [39]:
df1 = pd.read_csv(file_path, encoding="latin-1")

In [40]:
df1.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

         ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0   2/24/2003 0:00  Shipped       1         2     2003  ...   
1    5/7/2003 0:00  Shipped       2         5     2003  ...   
2    7/1/2003 0:00  Shipped       3         7     2003  ...   
3   8/25/2003 0:00  Shipped       3         8     2003  ...   
4  10/10/2003 0:00  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue           NaN            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA       NaN              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

In [41]:
df1.shape

(2823, 25)

In [42]:
df1.isnull().sum()

ORDERNUMBER            0
QUANTITYORDERED        0
PRICEEACH              0
ORDERLINENUMBER        0
SALES                  0
ORDERDATE              0
STATUS                 0
QTR_ID                 0
MONTH_ID               0
YEAR_ID                0
PRODUCTLINE            0
MSRP                   0
PRODUCTCODE            0
CUSTOMERNAME           0
PHONE                  0
ADDRESSLINE1           0
ADDRESSLINE2        2521
CITY                   0
STATE               1486
POSTALCODE            76
COUNTRY                0
TERRITORY           1074
CONTACTLASTNAME        0
CONTACTFIRSTNAME       0
DEALSIZE               0
dtype: int64

### Here in this dataset Out of 2823 data 2521 data has addressline 2 as null thus dropping the column along with Territory and state

### But in Postal code only 76 values are missing thus we will only remove those rows instead of fully deleting the column

In [43]:
df1.drop(['STATE','ADDRESSLINE2','TERRITORY'],axis = 1,inplace=True)

In [44]:
df1.shape

(2823, 22)

In [45]:
df1.dropna(inplace=True)

In [46]:
df1.shape

(2747, 22)

In [47]:
df1.isnull().sum()

ORDERNUMBER         0
QUANTITYORDERED     0
PRICEEACH           0
ORDERLINENUMBER     0
SALES               0
ORDERDATE           0
STATUS              0
QTR_ID              0
MONTH_ID            0
YEAR_ID             0
PRODUCTLINE         0
MSRP                0
PRODUCTCODE         0
CUSTOMERNAME        0
PHONE               0
ADDRESSLINE1        0
CITY                0
POSTALCODE          0
COUNTRY             0
CONTACTLASTNAME     0
CONTACTFIRSTNAME    0
DEALSIZE            0
dtype: int64

### Now the data is fully null value free. Thus we have handled the null values

In [49]:
df1.duplicated().sum()

0

### Dataset has no duplicate values

# Now finding the columns with string values then standardising them like making them consistent like in lowercase

In [53]:
string_columns = df1.select_dtypes(include=['object', 'string']).columns
print("Columns with string values:", list(string_columns))

Columns with string values: ['ORDERDATE', 'STATUS', 'PRODUCTLINE', 'PRODUCTCODE', 'CUSTOMERNAME', 'PHONE', 'ADDRESSLINE1', 'CITY', 'POSTALCODE', 'COUNTRY', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME', 'DEALSIZE']


In [54]:
df1[string_columns] = df1[string_columns].apply(lambda col: col.str.strip().str.lower())

In [55]:
df1

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
5           10168               36      96.66                1  3479.76   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

            ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ... PRODUCTCODE  \
0      2/24/2003 0:00   shipped       1         2     2003  ...    s10_1678   
1       5/7/2003 0:00   shipped       2         5     2003  ...    s10_1678   
2       7/1/2003 0:00   shipped       3         7     2003  ...    s10_1678   
3      8/25/2003 0:00   shipped       3         8     2003  ...    s10_1678   
5     10/28/2003 0:00   shipped       4        10     2003  ...    s10_1678   
...               ...       ...     ...       ...      ...  ...         ...   
2818   12/2/2004 0:00   shipped       4        12     2004  ...    s72_3212   
2819   1/31/2005 0:00   shipped       1         1     2005  ...    s72_3212   
2820    3/1/2005 0:00  resolved       1         3     2005  ...    s72_3212   
2821   3/28/2005 0:00   shipped       1         3     2005  ...    s72_3212   
2822    5/6/2005 0:00   on hold       2         5     2005  ...    s72_3212   

                 CUSTOMERNAME             PHONE  \
0           land of toys inc.        2125557818   
1          reims collectables        26.47.1555   
2             lyon souveniers  +33 1 46 62 7555   
3           toys4grownups.com        6265557265   
5        technics stores inc.        6505556809   
...                       ...               ...   
2818    euro shopping channel    (91) 555 94 44   
2819  oulu toy supplies, inc.        981-443655   
2820    euro shopping channel    (91) 555 94 44   
2821             alpha cognac        61.77.6555   
2822        gifts4allages.com        6175559555   

                       ADDRESSLINE1        CITY POSTALCODE  COUNTRY  \
0           897 long airport avenue         nyc      10022      usa   
1                59 rue de l'abbaye       reims      51100   france   
2     27 rue du colonel pierre avia       paris      75508   france   
3                78934 hillside dr.    pasadena      90003      usa   
5                 9408 furth circle  burlingame      94217      usa   
...                             ...         ...        ...      ...   
2818             c/ moralzarzal, 86      madrid      28034    spain   
2819                    torikatu 38        oulu      90110  finland   
2820             c/ moralzarzal, 86      madrid      28034    spain   
2821          1 rue alsace-lorraine    toulouse      31000   france   
2822             8616 spinnaker dr.      boston      51003      usa   

     CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0                 yu             kwai    small  
1            henriot             paul    small  
2           da cunha           daniel   medium  
3              young            julie   medium  
5             hirano             juri   medium  
...              ...              ...      ...  
2818          freyre            diego    small  
2819       koskitalo           pirkko   medium  
2820          freyre            diego   medium  
2821          roulet          annette    small  
2822         yoshido             juri   medium  

[2747 rows x 22 columns]

In [56]:
numeric_cols = df1.select_dtypes(include=['int','float']).columns

In [57]:
numeric_cols

Index(['ORDERNUMBER', 'QUANTITYORDERED', 'PRICEEACH', 'ORDERLINENUMBER',
       'SALES', 'QTR_ID', 'MONTH_ID', 'YEAR_ID', 'MSRP'],
      dtype='object')

In [59]:
len(numeric_cols)+ len(string_columns)

22

#### We have removed all the trailing spaces in string dtype columns

## Converting "ORDERDATE" column to datetime 

In [69]:
df1['ORDERDATE'] = pd.to_datetime(df1['ORDERDATE'])

In [70]:
df1

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
5           10168               36      96.66                1  3479.76   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

      ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ... PRODUCTCODE  \
0    2003-02-24   shipped       1         2     2003  ...    s10_1678   
1    2003-05-07   shipped       2         5     2003  ...    s10_1678   
2    2003-07-01   shipped       3         7     2003  ...    s10_1678   
3    2003-08-25   shipped       3         8     2003  ...    s10_1678   
5    2003-10-28   shipped       4        10     2003  ...    s10_1678   
...         ...       ...     ...       ...      ...  ...         ...   
2818 2004-12-02   shipped       4        12     2004  ...    s72_3212   
2819 2005-01-31   shipped       1         1     2005  ...    s72_3212   
2820 2005-03-01  resolved       1         3     2005  ...    s72_3212   
2821 2005-03-28   shipped       1         3     2005  ...    s72_3212   
2822 2005-05-06   on hold       2         5     2005  ...    s72_3212   

                 CUSTOMERNAME             PHONE  \
0           land of toys inc.        2125557818   
1          reims collectables        26.47.1555   
2             lyon souveniers  +33 1 46 62 7555   
3           toys4grownups.com        6265557265   
5        technics stores inc.        6505556809   
...                       ...               ...   
2818    euro shopping channel    (91) 555 94 44   
2819  oulu toy supplies, inc.        981-443655   
2820    euro shopping channel    (91) 555 94 44   
2821             alpha cognac        61.77.6555   
2822        gifts4allages.com        6175559555   

                       ADDRESSLINE1        CITY POSTALCODE  COUNTRY  \
0           897 long airport avenue         nyc      10022      usa   
1                59 rue de l'abbaye       reims      51100   france   
2     27 rue du colonel pierre avia       paris      75508   france   
3                78934 hillside dr.    pasadena      90003      usa   
5                 9408 furth circle  burlingame      94217      usa   
...                             ...         ...        ...      ...   
2818             c/ moralzarzal, 86      madrid      28034    spain   
2819                    torikatu 38        oulu      90110  finland   
2820             c/ moralzarzal, 86      madrid      28034    spain   
2821          1 rue alsace-lorraine    toulouse      31000   france   
2822             8616 spinnaker dr.      boston      51003      usa   

     CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0                 yu             kwai    small  
1            henriot             paul    small  
2           da cunha           daniel   medium  
3              young            julie   medium  
5             hirano             juri   medium  
...              ...              ...      ...  
2818          freyre            diego    small  
2819       koskitalo           pirkko   medium  
2820          freyre            diego   medium  
2821          roulet          annette    small  
2822         yoshido             juri   medium  

[2747 rows x 22 columns]

## Renaming column headers to be clean and uniform (e.g., lowercase, no spaces)

In [71]:
df1.columns = [col.strip().lower().replace(" ", "_") for col in df1.columns]

In [72]:
df1.columns

Index(['ordernumber', 'quantityordered', 'priceeach', 'orderlinenumber',
       'sales', 'orderdate', 'status', 'qtr_id', 'month_id', 'year_id',
       'productline', 'msrp', 'productcode', 'customername', 'phone',
       'addressline1', 'city', 'postalcode', 'country', 'contactlastname',
       'contactfirstname', 'dealsize'],
      dtype='object')

### Checking and fixing data types (e.g., age should be int, date as datetime).

In [75]:
int_cols = df1.select_dtypes(include=['int']).columns
float_cols = df1.select_dtypes(include=['float']).columns
string_cols = df1.select_dtypes(include=['string']).columns
datetime_cols = df1.select_dtypes(include=['datetime']).columns

In [76]:
print(f"int Columns = {int_cols}\nfloat columns = {float_cols}\nstring columns = {string_cols}\ndatetime columns = {datetime_cols}")

int Columns = Index(['ordernumber', 'quantityordered', 'orderlinenumber', 'qtr_id',
       'month_id', 'year_id', 'msrp'],
      dtype='object')
float columns = Index(['priceeach', 'sales'], dtype='object')
string columns = Index([], dtype='object')
datetime columns = Index(['orderdate'], dtype='object')


## Now we have successfully converted each columns into its original datatype